### PR TITLE
add reset-event and reset-goal

### PR DESCRIPTION
**Pull request by Goober5000**
These new SEXP operators clear all status information from an event or goal, making it as if the event or goal had never been evaluated.

Implements feature discussed in #4184.

~~In draft status pending tests.~~  Now tested.  ~~Depends on #4633.~~

### CODE REVIEW

File: code/parse/sexp.cpp [hunk 1]
```diff
@@ -17367,6 +17369,59 @@ int sexp_goal_incomplete(int n)
...
+		auto name = CTEXT(n);
+		int event_num = mission_event_lookup(name);
+		if (event_num >= 0)
+		{
+			auto eventp = &Mission_events[event_num];
+
+			eventp->result = 0;
+			eventp->timestamp = TIMESTAMP::invalid();
+
+			// clear some flags but not all
```
Comment (naomimyselfandi) [inline code/parse/sexp.cpp:17388]: I don't know if this comment adds any value. It just says what the code does, and there isn't any complex magic here that benefits from being summarized. The comment would be more useful if it said why it's clearing these flags in particular.

File: code/parse/sexp.cpp [hunk 2]
```diff
@@ -17367,6 +17369,59 @@ int sexp_goal_incomplete(int n)
...
+			eventp->flags &= ~MEF_DIRECTIVE_SPECIAL;
+			eventp->flags &= ~MEF_DIRECTIVE_TEMP_TRUE;
+			eventp->flags &= ~MEF_TIMESTAMP_HAS_INTERVAL;
+
+			eventp->count = 0;
+			eventp->satisfied_time = TIMESTAMP::invalid();
+			eventp->born_on_date = TIMESTAMP::invalid();
+			eventp->previous_result = 0;
+
+			if (eventp->formula >= 0)
```
Comment (naomimyselfandi) [inline code/parse/sexp.cpp:17399]: `flush_sexp_tree` already short-circuits on a negative node index.

(More curiosity than anything, but when would an event not have a formula?)

File: code/parse/sexp.cpp [hunk 3]
```diff
@@ -17367,6 +17369,59 @@ int sexp_goal_incomplete(int n)
...
+				flush_sexp_tree(eventp->formula);
+		}
```
Comment (naomimyselfandi) [inline code/parse/sexp.cpp:17401]: This should probably warn if the lookup failed.

File: code/parse/sexp.cpp [hunk 4]
```diff
@@ -17367,6 +17369,59 @@ int sexp_goal_incomplete(int n)
...
+	}
+}
+
+/**
+ * Resets a goal, its status, and its nodes so that it is as if the goal had never been evaluated.
+ */
+void sexp_reset_goal(int node)
```
Comment (naomimyselfandi) [inline code/parse/sexp.cpp:17408]: Same comments as above about `flush_sexp_tree` and warning on lookup failure.

**Comment by Goober5000** [comment]
Just a note to say thanks for the review.  I'll get to these comments as soon as I can - just been busy.

File: code/parse/sexp.cpp (hunk 1 above)
Comment (Goober5000) [reply code/parse/sexp.cpp]: Good point

File: code/parse/sexp.cpp (hunk 2 above)
Comment (Goober5000) [reply code/parse/sexp.cpp]: So it does.  Tweaked.

Hmm... it looks like an event's formula is set to -1 when it is failed, or known-false.  That will need to be modified for this PR to work.

File: code/parse/sexp.cpp (hunk 3 above)
Comment (Goober5000) [reply code/parse/sexp.cpp]: Good point

Comment (Goober5000) [reply code/parse/sexp.cpp]: Actually... none of the other sexps warn when an event lookup fails.  This includes the retail `is-event-true-delay` etc.  I suppose that shouldn't rule out warning here though.

**Comment by Goober5000** [comment]
Comments addressed.  I'll still need to figure out a way to prevent the event formula from being set to -1.

**Comment by Goober5000** [comment]
Ok, this has been tested and should be ready now.

**Comment by JohnAFernandez** [comment]
I took a look, and I'm ready to approve pending another quick look by @naomimyselfandi

**Pull request merged**